app/services/usps_in_person_proofer.rb

```diff
@@ -167,24 +93,79 @@ def request_enrollment_code(unique_id)
     body = {
       sponsorID: sponsor_id,
       uniqueID: unique_id,
-    }.to_json
+    }
+
+    faraday.post(url, body, dynamic_headers).body
+  end
+
+  # Makes a request to retrieve a new OAuth token
+  # and modifies self to store the token and when
+  # it expires (15 minutes).
+  # @return [String] the token
+  def retrieve_token!
+    body = request_token
+    @token_expires_at = Time.zone.now + body['expires_in']
+    @token = "#{body['token_type']} #{body['access_token']}"
+  end
+
+  def token_valid?
+    @token.present? && @token_expires_at.present? && @token_expires_at.future?
+  end
+
+  private
 
-    headers = request_headers.merge(
-      {
-        'Authorization' => @token,
-        'RequestID' => request_id,
-      },
-    )
+  def faraday
+    Faraday.new(headers: request_headers) do |conn|
+      conn.options.timeout = IdentityConfig.store.usps_ipp_request_timeout
+      conn.options.read_timeout = IdentityConfig.store.usps_ipp_request_timeout
+      conn.options.open_timeout = IdentityConfig.store.usps_ipp_request_timeout
+      conn.options.write_timeout = IdentityConfig.store.usps_ipp_request_timeout
+
+      # Raise an error subclassing Faraday::Error on 4xx, 5xx, and malformed responses
+      # Note: The order of this matters for parsing the error response body.
+      conn.response :raise_error
 
-    resp = faraday.post(url, body, headers)
+      # Log request method and URL, excluding headers and body
+      conn.response :logger, nil, { headers: false, bodies: false }
 
-    if resp.success?
-      JSON.parse(resp.body)
-    else
-      resp
+      # Convert body to JSON
+      conn.request :json
+
+      # Parse JSON responses
+      conn.response :json
     end
   end
 
+  # Retrieve the OAuth2 token (if needed) and then pass
+  # the headers to an arbitrary block of code as a Hash.
+  #
+  # Returns the same value returned by that block of code.
+  def dynamic_headers
+    retrieve_token! unless token_valid?
+
+    {
+      'Authorization' => @token,
+      'RequestID' => request_id,
+    }
+  end
+
+  # Makes HTTP request to authentication endpoint and
+  # returns the token and when it expires (15 minutes).
+  # @return [Hash] API response
+  def request_token
+    url = "#{root_url}/oauth/authenticate"
+    body = {
+      username: IdentityConfig.store.usps_ipp_username,
+      password: IdentityConfig.store.usps_ipp_password,
+      grant_type: 'implicit',
+      response_type: 'token',
+      client_id: '424ada78-62ae-4c53-8e3a-0b737708a9db',
```

Comment:
I'm not familiar with this and I see that it existed previously, but should this value be stored in configuration?

Reply:
When it was added [we decided not to put it in config](https://github.com/18F/identity-idp/pull/4523#discussion_r546927755), but I think it makes sense to do so. I'll create a new ticket for that.

Reply:
ticket LG-6925 created